crates/forge/src/block_number_map.rs

```diff
@@ -0,0 +1,118 @@
+use crate::compiled_raw::RawForkParams;
+use anyhow::{anyhow, Result};
+use cairo_felt::Felt252;
+use conversions::IntoConv;
+use forge_runner::compiled_runnable::ValidatedForkConfig;
+use num_bigint::BigInt;
+use starknet::core::types::BlockTag::Latest;
+use starknet::core::types::{BlockId, MaybePendingBlockWithTxHashes};
+use starknet::providers::jsonrpc::HttpTransport;
+use starknet::providers::{JsonRpcClient, Provider};
+use starknet_api::block::BlockNumber;
+use std::collections::HashMap;
+use tokio::runtime::Handle;
+use url::Url;
+
+#[derive(Default)]
+pub struct BlockNumberMap {
+    url_to_latest_block_number: HashMap<String, BlockNumber>,
+    url_and_hash_to_block_number: HashMap<(String, Felt252), BlockNumber>,
+}
+
+impl BlockNumberMap {
+    pub fn add_latest_block_number(&mut self, url: String, latest_block_number: BlockNumber) {
+        self.url_to_latest_block_number
+            .insert(url, latest_block_number);
+    }
+
+    pub fn add_block_number_for_hash(
+        &mut self,
+        url: String,
+        hash: Felt252,
+        block_number: BlockNumber,
+    ) {
+        self.url_and_hash_to_block_number
+            .insert((url, hash), block_number);
+    }
+
+    pub fn get_latest_block_number(&mut self, url: &String) -> Option<&BlockNumber> {
+        self.url_to_latest_block_number.get(url)
+    }
+
+    pub fn get_block_number_for_hash(
+        &mut self,
+        url: String,
+        hash: Felt252,
+    ) -> Option<&BlockNumber> {
+        self.url_and_hash_to_block_number.get(&(url, hash))
+    }
+
+    #[must_use]
+    pub fn get_url_to_latest_block_number(&self) -> &HashMap<String, BlockNumber> {
+        &self.url_to_latest_block_number
+    }
+}
+
+async fn get_latest_block_number(url: &Url) -> Result<BlockNumber> {
+    let client = JsonRpcClient::new(HttpTransport::new(url.clone()));
+
+    match Handle::current()
+        .spawn(async move { client.get_block_with_tx_hashes(BlockId::Tag(Latest)).await })
```

Comment:
Btw this is the solution that should be used to solve https://github.com/foundry-rs/starknet-foundry/issues/901
cc @cptartur @tkumor3 @Arcticae @MaksymilianDemitraszek